src/locale/keymap-util.c

```diff
@@ -780,3 +782,141 @@ int x11_convert_to_vconsole(Context *c) {
+        } else
+                return -EINVAL; /* We do not process non-UTF-8 locale */
+
+        fr = fopen("/etc/locale.gen", "re");
+        if (!fr)
+                return -errno;
+
+        r = fopen_temporary("/etc/locale.gen", &fw, &temp_path);
+        if (r < 0)
+                return r;
+
+        /* apply mode of the original file to new file */
+        if (stat("/etc/locale.gen", &st) < 0)
+                return -errno;
+        (void) fchmod(fileno(fw), st.st_mode);
+
+        /* The config file ends with a line break, which we do not want to include before potentially appending a new locale
+         * instead of uncommenting an existing line. By prepending linebreaks, we can avoid buffering this file but can still write
+         * a nice config file without empty lines */
+        first_line = true;
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                char *line_locale;
+
+                r = read_line(fr, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (locale_enabled) {
+                        /* Just complete writing the file if the new locale was already enabled */
+                        fprintf(fw, "%s%s", first_line? "" : "\n", line);
+                        first_line = false;
+                        continue;
+                }
+
+                line = strstrip(line);
+                if (isempty(line)) {
+                        fputs("\n", fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line_locale = line;
+                if (line_locale[0] == '#')
+                        line_locale = strstrip(line_locale + 1);
+
+                if (strcasecmp_ptr(line_locale, locale_entry) == 0) {
+                        /* Uncomment existing line for new locale */
+                        fprintf(fw, "%s%s", first_line? "" : "\n", locale_entry);
+                        locale_enabled = true;
+                        first_line = false;
+                        continue;
+                }
+
+                /* The line was not for the locale we want to enable, just copy it */
+                fprintf(fw, "%s%s", first_line? "" : "\n", line);
+                first_line = false;
+        }
+
+        /* Add locale to enable to the end of the file if it was not found as commented line */
+        if (!locale_enabled)
+                fprintf(fw, "\n%s", locale_entry);
+        fputs("\n", fw);
+
+        r = fflush_sync_and_check(fw);
+        if (r < 0)
+                goto fail;
+
+        if (rename(temp_path, "/etc/locale.gen") < 0) {
+                r = -errno;
+                goto fail;
+        }
+        temp_path = mfree(temp_path);
+
+        return 0;
+
+fail:
+        if (temp_path)
+                (void) unlink(temp_path);
+        return r;
```

Comment:
you can drop these four lines altogether. The `_cleanup_(unlink_and_freep)` cleanup stuff you now use does this anyway whenever the code block is left and temp_path is not NULL. Thus whenever you see an error you can just do "return r" directly, without any goto whatsoever.